standards/application/relay-reliability.md

```diff
@@ -0,0 +1,192 @@
+
+The implementation in Python may look like this:
+
+```python
+outgoingMessageHashes = []
+
+class Message:
+    hash: str
+    postTime: int
+    status: str
+    content: str
+
+def send(message):
+    # send message via relay or lightpush protocol, here use relay as example
+    waku.relay.post(message)
+    outgoingMessageHashes.append(message.hash)
+
+    message.status = 'outgoing'
+    database.saveMessage(message)
+
+def checkOutgoingMessages(peerID):
+    for messageHash in outgoingMessageHashes:
+        message = database.getMessage(messageHash)
+        # only query store node for ongoing message, and posted more than 3 seconds ago
+        if message.status == 'ongoing' && time.now() - message.postTime > 3:
+            response = waku.store.queryMessage(peerID, messageHash)
+            if response.exists():
+                database.updateMessageStatus(messageHash, 'sent')
+                outgoingMessageHashes.remove(messageHash)
+            elif time.now() - message.postTime > 10:
+                # resend the message if it's not stored in store node after 10 seconds
+                waku.relay.post(message)
+```
+
+Function `checkOutgoingMessages` is called periodically, most likely every a few seconds. Message hashes can be queried in batch to reduce the number of requests to store nodes, the size in a batch shoud not exceed the max supported size by store node.
+
+The store node can be set and updated directly by application or selected from peers which are discovery by protocols like [discv5](https://github.com/vacp2p/rfc-index/blob/main/waku/standards/core/33/discv5.md) or [peer exchange](https://github.com/waku-org/specs/blob/master/standards/core/peer-exchange.md).
+
+The store node may only support specific pubsub topics, and the application should group message hashes by pubsub topic before sending the request.
+
+When persistent network issue happens, you may not want to resend the failed messages indefinitely, the application should have a mechanism to clean the cache with failed message hashes and trigger other retry logic after a few attempts.
+
+### Query with Topics and Time Range
+
+An application could use different pubsub topics and content topics, for example a community may have its own pubsub topic, and each channel may have its own content topic. To fetch all missing messages in a specific channel, the application can query the store node with the provided pubsub topic, content topic and time range.
+
+For incoming messages, the processing flow can be like this:
+- subscribe to the interested pubsub and content topics
+- query the store node with the interested topics and time range for message hashes periodically
+- check if the message hash is exist in local database, if not, adding the message hash to a buffer, if yes, skip the message
+- fetch the missing messages in the buffer from store node
+- process the messages necessarily
+- update the last fetch time for the interested topic
```

Comment:
I think this can be removed from the heuristic if we just make it clear that the heuristic must be run periodically for all content topics interesting to the appplication.